src/main/cpp/src/slice.cu

```diff
@@ -0,0 +1,358 @@
+  }
+}
+
+template <typename SRART_ITERATOR, typename LENGTH_ITERATOR>
+auto generate_starts_and_sizes(size_type const* offsets_of_input_lists,
+                               size_type const num_rows,
+                               SRART_ITERATOR const start_iterator,
+                               LENGTH_ITERATOR const length_iterator,
+                               rmm::cuda_stream_view stream)
+{
+  auto starts              = make_numeric_column(data_type{type_id::INT32},
+                                    num_rows,
+                                    mask_state::UNALLOCATED,
+                                    stream,
+                                    cudf::get_current_device_resource_ref());
+  auto sizes               = make_numeric_column(data_type{type_id::INT32},
+                                   num_rows,
+                                   mask_state::UNALLOCATED,
+                                   stream,
+                                   cudf::get_current_device_resource_ref());
+  constexpr int block_size = 256;
+  auto grid                = cudf::detail::grid_1d{num_rows, block_size};
+  compute_starts_and_sizes_kernel<<<grid.num_blocks, block_size, 0, stream.value()>>>(
+    offsets_of_input_lists,
+    num_rows,
+    start_iterator,
+    length_iterator,
+    starts->mutable_view().data<int32_t>(),
+    sizes->mutable_view().data<int32_t>());
+  return std::make_pair(std::move(starts), std::move(sizes));
+}
+
+std::unique_ptr<cudf::column> legal_slice(lists_column_view const& input,
+                                          column_view const& starts,
+                                          column_view const& sizes,
+                                          rmm::cuda_stream_view stream,
+                                          rmm::device_async_resource_ref mr)
+{
+  auto const num_rows = input.size();
+  // make output_offset
+  auto [output_offset, num_total_elements] = cudf::detail::make_offsets_child_column(
+    sizes.begin<int32_t>(), sizes.end<int32_t>(), stream, mr);
+
+  // generate gather map
+  constexpr int block_size = 256;
+  auto grid                = cudf::detail::grid_1d{num_rows, block_size};
+  rmm::device_uvector<int32_t> gather_map(num_total_elements, stream);
+  compute_gather_map<<<grid.num_blocks, block_size, 0, stream.value()>>>(
+    num_rows,
+    input.offsets_begin(),
+    starts.begin<int32_t>(),
+    output_offset->view().begin<int32_t>(),
+    gather_map.data());
+
+  // Call gather on child of input column
+  auto child_table = cudf::detail::gather(table_view({input.get_sliced_child(stream)}),
+                                          gather_map,
+                                          out_of_bounds_policy::DONT_CHECK,
+                                          cudf::detail::negative_index_policy::NOT_ALLOWED,
+                                          stream,
+                                          cudf::get_current_device_resource_ref());
```

Comment:
```suggestion
                                          mr
```

The `child` column is a part of output result, so use the `mr` passed in from API.
For the internal temp virable, we can use `cudf::get_current_device_resource_ref`.
I remember Nigha had mentioned this.